### PR TITLE
VillageState Sync

### PR DESCRIPTION
**Pull request by EgardA**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] All new classes have class-level documentation comments, if there are any at all
- [ ] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation update
- [ ] Other... Please describe:

## What is the new behaviour?

Resolves #

## Other information

### CODE REVIEW

File: source/GameInterface/Services/Villages/Patches/ChangeVillageStatePatch.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+{
+    [HarmonyPatch(typeof(ChangeVillageStateAction), "ApplyInternal")]
+    public class ChangeVillageStatePatch
+    {
+        private static AllowedInstance<Village> _allowedInstance;
+        private static readonly MethodInfo _applyInternal = typeof(ChangeVillageStateAction).GetMethod("ApplyInternal", BindingFlags.NonPublic | BindingFlags.Static);
+
+        public static bool Prefix(Village village, Village.VillageStates newState, MobileParty raiderParty)
+        {
+            if (village == _allowedInstance?.Instance) return true;
```
Comment (garrettluskey) [inline source/GameInterface/Services/Villages/Patches/ChangeVillageStatePatch.cs]: We now have a method for checking if allowed

```suggestion
            if (_allowedInstance.IsAllowed(village)) return true;
```

File: source/GameInterface/Services/Villages/Patches/ChangeVillageStatePatch.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+using TaleWorlds.CampaignSystem.Party;
+using TaleWorlds.CampaignSystem.Settlements;
+
+namespace GameInterface.Services.Villages.Patches
+{
+    [HarmonyPatch(typeof(ChangeVillageStateAction), "ApplyInternal")]
+    public class ChangeVillageStatePatch
+    {
+        private static AllowedInstance<Village> _allowedInstance;
+        private static readonly MethodInfo _applyInternal = typeof(ChangeVillageStateAction).GetMethod("ApplyInternal", BindingFlags.NonPublic | BindingFlags.Static);
```
Comment (garrettluskey) [inline source/GameInterface/Services/Villages/Patches/ChangeVillageStatePatch.cs]: I don't remember if this works with statics but give it a try. If not let me know and keep as is.

I want any calls that happen a lot to use delegates as reflection invoke is fairly slow

```suggestion
        private static MethodInfo _applyInternal => typeof(ChangeVillageStateAction).GetMethod("ApplyInternal", BindingFlags.NonPublic | BindingFlags.Static);
        private static readonly Action<Village, Village.VillageStates, MobileParty> ApplyInternalDelegate = _applyInternal .BuildDelegate<Action<Village, Village.VillageStates, MobileParty>>();
```

File: source/GameInterface/Services/Villages/Messages/VillageStateChanged.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿using Common.Messaging;
+
+namespace GameInterface.Services.Villages.Messages
+{
+    public record VillageStateChanged : IEvent
```
Comment (garrettluskey) [inline source/GameInterface/Services/Villages/Messages/VillageStateChanged.cs:8]: Please add summary comments

File: source/GameInterface/Services/Villages/Messages/ChangeVillageStateRequest.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
+﻿using Common.Messaging;
+using ProtoBuf;
+
+namespace GameInterface.Services.Villages.Messages
+{
+    [ProtoContract(SkipConstructor = true)]
+    public record ChangeVillageStateRequest : ICommand
```
Comment (garrettluskey) [inline source/GameInterface/Services/Villages/Messages/ChangeVillageStateRequest.cs]: Please add summary comments

File: source/GameInterface/Services/Villages/Messages/ChangeVillageStateApproved.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+﻿using Common.Messaging;
+using ProtoBuf;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Coop.Core.Client.Messages
+{
+    [ProtoContract(SkipConstructor = true)]
+    public record ChangeVillageStateApproved : ICommand
```
Comment (garrettluskey) [inline source/GameInterface/Services/Villages/Messages/ChangeVillageStateApproved.cs]: Please add summary comments

File: source/GameInterface/Services/Villages/Messages/ChangeVillageState.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
+﻿using Common.Messaging;
+
+namespace GameInterface.Services.Villages.Messages
+{
+    public record ChangeVillageState : IEvent
```
Comment (garrettluskey) [inline source/GameInterface/Services/Villages/Messages/ChangeVillageState.cs:8]: Please add summary comments

File: source/Coop.Core/Client/Messages/ChangeVillageStateApproved.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
+﻿using Common.Messaging;
+using ProtoBuf;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Coop.Core.Client.Messages
+{
+    [ProtoContract(SkipConstructor = true)]
+    public record ChangeVillageStateApproved : ICommand
```
Comment (garrettluskey) [inline source/Coop.Core/Client/Messages/ChangeVillageStateApproved.cs:10]: Please add summary comments

File: source/GameInterface/Services/Villages/Messages/ChangeVillageStateRequest.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
+﻿using Common.Messaging;
+using ProtoBuf;
+
+namespace GameInterface.Services.Villages.Messages
```
Comment (garrettluskey) [inline source/GameInterface/Services/Villages/Messages/ChangeVillageStateRequest.cs]: This shouldn't be in game interface

File: source/GameInterface/Services/Villages/Messages/ChangeVillageStateApproved.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
+﻿using Common.Messaging;
+using ProtoBuf;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Coop.Core.Client.Messages
```
Comment (garrettluskey) [inline source/GameInterface/Services/Villages/Messages/ChangeVillageStateApproved.cs]: This shouldn't be in game interface

File: source/GameInterface/Services/Villages/Patches/ChangeVillageStatePatch.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+using TaleWorlds.CampaignSystem.Actions;
+using TaleWorlds.CampaignSystem.Party;
+using TaleWorlds.CampaignSystem.Settlements;
+
+namespace GameInterface.Services.Villages.Patches
+{
+    [HarmonyPatch(typeof(ChangeVillageStateAction), "ApplyInternal")]
+    public class ChangeVillageStatePatch
+    {
+        private static AllowedInstance<Village> _allowedInstance;
```
Comment (garrettluskey) [inline source/GameInterface/Services/Villages/Patches/ChangeVillageStatePatch.cs]:
```suggestion
        private readonly static AllowedInstance<Village> _allowedInstance = new AllowedInstance<Village>();
```

File: source/GameInterface/Services/Villages/Patches/ChangeVillageStatePatch.cs [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+            if (village == _allowedInstance?.Instance) return true;
+
+            MessageBroker.Instance.Publish(village, new ChangeVillageState(village.StringId, Convert.ToInt32(newState), raiderParty.StringId));
+
+            return false;
+        }
+
+        public static void RunOriginalApplyInternal(Village village, Village.VillageStates newState, MobileParty raiderParty)
+        {
+            using (_allowedInstance = new AllowedInstance<Village>(village))
```
Comment (garrettluskey) [inline source/GameInterface/Services/Villages/Patches/ChangeVillageStatePatch.cs]:
```suggestion
            using (_allowedInstance.Instance = village)
```

**Review by garrettluskey — CHANGES_REQUESTED** [review]

File: source/GameInterface/Services/Villages/Handlers/VillageHandler.cs [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+        private void Handle(MessagePayload<VillageStateChanged> obj)
+        {
+            if (objectManager.TryGetObject(obj.What.PartyId, out MobileParty mobileParty) == false)
+            {
+                Logger.Error("Could not handle VillageStateChanged, PartyId not found: {id}", obj.What.PartyId);
+                return;
+            }
+
+            Village village = Village.All.Find(x => x.StringId == obj.What.VillageId);
```
Comment (garrettluskey) [inline source/GameInterface/Services/Villages/Handlers/VillageHandler.cs:49]: This is fairly slow, can you use objectManager instead? Settlements should all be registered

File: source/GameInterface/Services/Villages/Patches/ChangeVillageStatePatch.cs [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+            if (raiderParty == null) return false;
+
+            MessageBroker.Instance.Publish(village, new ChangeVillageState(village.StringId, Convert.ToInt32(newState), raiderParty.StringId));
+
+            return false;
+        }
+
+        public static void RunOriginalApplyInternal(Village village, Village.VillageStates newState, MobileParty raiderParty)
+        {
+            using (_allowedInstance)
```
Comment (garrettluskey) [inline source/GameInterface/Services/Villages/Patches/ChangeVillageStatePatch.cs:35]: Update allowed instance to AllowedThread after https://github.com/Bannerlord-Coop-Team/BannerlordCoop/pull/723 is merged

File: source/GameInterface/Services/Villages/Patches/ChangeVillageStatePatch.cs [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+        private static MethodInfo _applyInternal => typeof(ChangeVillageStateAction).GetMethod("ApplyInternal", BindingFlags.NonPublic | BindingFlags.Static);
+        private static readonly Action<Village, Village.VillageStates, MobileParty> ApplyInternalDelegate = _applyInternal.BuildDelegate<Action<Village, Village.VillageStates, MobileParty>>();
+
+        public static bool Prefix(Village village, Village.VillageStates newState, MobileParty raiderParty)
+        {
+            if (_allowedInstance.IsAllowed(village)) return true;
+
+            if (raiderParty == null) return false;
+
+            MessageBroker.Instance.Publish(village, new ChangeVillageState(village.StringId, Convert.ToInt32(newState), raiderParty.StringId));
```
Comment (garrettluskey) [inline source/GameInterface/Services/Villages/Patches/ChangeVillageStatePatch.cs:28]: This event should be `VillageStateChanged`

File: source/GameInterface/Services/Villages/Patches/ChangeVillageStatePatch.cs [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+using HarmonyLib;
+using System;
+using System.Reflection;
+using TaleWorlds.CampaignSystem.Actions;
+using TaleWorlds.CampaignSystem.Party;
+using TaleWorlds.CampaignSystem.Settlements;
+
+namespace GameInterface.Services.Villages.Patches
+{
+    [HarmonyPatch(typeof(ChangeVillageStateAction), "ApplyInternal")]
```
Comment (garrettluskey) [inline source/GameInterface/Services/Villages/Patches/ChangeVillageStatePatch.cs:15]: Instead sync `VillageState` setter

You can keep this patch but update prefix to `private static bool Prefix() => ModInformation.IsServer`

File: source/Coop.Core/Client/Services/Villages/Messages/ChangeVillageStateRequest.cs [hunk 16]
```diff
@@ -0,0 +1,26 @@
+﻿using Common.Messaging;
+using ProtoBuf;
+
+namespace Coop.Core.Client.Services.Villages.Messages
+{
+    /// <summary>
+    /// Request to change village state from client to server
+    /// </summary>
+    [ProtoContract(SkipConstructor = true)]
+    public record ChangeVillageStateRequest : ICommand
```
Comment (garrettluskey) [inline source/Coop.Core/Client/Services/Villages/Messages/ChangeVillageStateRequest.cs:10]: This should be server side only, no need for client requests

**Review by garrettluskey — CHANGES_REQUESTED** [review]

**Comment by EgardA** [comment]
Closed, as this would instead be a part of MapEvent as syncing something that only happens as part of something further up a call seems pointless

**Pull request closed without merging**